sympy/polys/agca/extensions.py

```diff
@@ -224,22 +240,53 @@ def __init__(self, mod):
         self.one = self.convert(self.ring.one)
 
         gen = self.ring.gens[0]
+        self.symbol = self.ring.symbols[0]
         self.generator = self.convert(gen)
-        self.basis = tuple(self.convert(gen**i)
-                      for i in range(self.rank))
+        self.basis = tuple(self.convert(gen**i) for i in range(self.rank))
 
-    def convert(self, f):
-        rep = self.ring.convert(f)
+    def new(self, arg):
+        rep = self.ring.convert(arg)
         return ExtElem(rep % self.mod, self)
 
-    def to_sympy(self, f):
-        return self.ring.to_sympy(f.rep)
+    def __eq__(self, other):
+        if not isinstance(other, FiniteExtension):
+            return False
+        return self.modulus == other.modulus
 
-    __call__ = convert
+    def __hash__(self):
+        return hash((self.__class__.__name__, self.modulus))
 
     def __str__(self):
         return "%s/(%s)" % (self.ring, self.modulus.as_expr())
 
     __repr__ = __str__
 
+    def convert(self, f, base=None):
+        rep = self.ring.convert(f, base)
+        return ExtElem(rep % self.mod, self)
+
+    def convert_from(self, f, base):
+        rep = self.ring.convert(f, base)
+        return ExtElem(rep % self.mod, self)
+
+    def to_sympy(self, f):
+        return self.ring.to_sympy(f.rep)
+
+    def from_sympy(self, f):
+        return self.convert(f)
+
+    def set_domain(self, K):
+        mod = self.modulus.set_domain(K)
+        return self.__class__(mod)
+
+    def drop(self, *symbols):
+        if self.symbol in symbols:
+            raise GeneratorsError('Can not drop generator from FiniteExtension')
+        K = self.domain.drop(*symbols)
+        return self.set_domain(K)
+
+    def exquo(self, f, g):
+        rep = self.domain.exquo(f.rep, g.rep)
+        return ExtElem(rep % self.mod, self)
```

Comment:
I'm not sure how to implement `exquo` here.

If self is `QQ[x]/(x**2 - 2)` then `self.domain` is `QQ`. At least this should call `self.ring.exquo` rather than `self.domain.exquo` (since the arguments are elements of `self.ring` so that's just a mistake). However `self.ring` is a ring whereas `self` is a field in which exact division should be possible for any nonzero divisor.

Reply:
Shouldn't this be the same as `__truediv__`? So this could be `return f / g`.

Reply:
If `f / g` is exact then does that not imply that `f // g == f / g` and `f % g == 0`?

Reply:
It looks like that is how `f // g` and `f % g` should be implemented when `ring.domain` is a field. Otherwise some kind of error should probably be returned.

Reply:
In a nested extension the base domain is also a finite extension:
```python
In [17]: from sympy.polys.agca.extensions import FiniteExtension

In [18]: K = FiniteExtension(Poly(x**2 - 2, field=True))

In [19]: K2 = FiniteExtension(Poly(y**2 - 2, domain=K))

In [20]: K2
Out[20]: QQ[x]/(x**2 - 2)[y]/(y**2 - 2)

In [21]: K2.domain
Out[21]: QQ[x]/(x**2 - 2)
```
We need to be able to determine whether or not this is a field.

Can we say that if `K` is a field and `p(x)` is irreducible in `K[x]` then `K[x]/p(x)` is a field?

Conversely can we say that if `K` is not a field or `p(x)` is reducible over `K[x]` then `K[x]/p(x)` is not a field?

It seems like we can do
```
    self.is_Field = domain.is_field and mod.is_irreducible
```
in `Domain.__init__`.

Then if `is_Field` is true we can allow full division with `/`, `//` and `%` as:
```
f / g  ==  f * g.invert(mod)
f // g == f / g
f % g == 0
```
Otherwise if `is_Field` is false then we can allow the same but only if g is in K (`g.is_ground`)?

In non-field bug ground case the base domain `K` can decide if the division should succeed or fail.

Reply:
It looks like `is_Field` currently means that "exact" division (by nonzerodivisors) is possible. It is true for RR though it is not really a field (only few elements are actually invertible). I would consider making `is_Field` true for `K[x]/p(x)` whenever `K` is true even if `p(x)` is not irreducible. If there turn out to be issues, then maybe `is_Field` could be None instead of false.

If `K` is not a field, then `is_Field` should be false. In that case, I think that it is safe to divide only by invertible constants g in K.

Reply:
Okay, I'll try that.

It seems that the `is_unit` method is broken for `ZZ` and `QQ[x]` unless I misunderstand:
```python
In [62]: ZZ.is_unit(ZZ(-1))
Out[62]: False

In [63]: K = QQ[x]

In [64]: K.is_unit(K(2))
Out[64]: False

In [66]: QQ.is_unit(QQ(2))
Out[66]: True

In [68]: QQ.is_unit(QQ(0))
Out[68]: False
```